src/core/ext/filters/client_channel/lb_policy/weighted_round_robin/weighted_round_robin.cc

```diff
@@ -0,0 +1,942 @@
+
+    RefCountedPtr<WeightedRoundRobin> wrr_;
+    const bool use_per_rpc_utilization_;
+    const Duration weight_update_period_;
+    const Duration weight_expiration_period_;
+    const Duration blackout_period_;
+    std::vector<SubchannelInfo> subchannels_;
+
+    Mutex mu_;
+    uint32_t scheduler_state_ ABSL_GUARDED_BY(&mu_);
+    absl::optional<StaticStrideScheduler> scheduler_ ABSL_GUARDED_BY(&mu_);
+    absl::optional<grpc_event_engine::experimental::EventEngine::TaskHandle>
+        timer_handle_ ABSL_GUARDED_BY(&mu_);
+
+    // Used when falling back to RR.
+    size_t last_picked_index_;
+  };
+
+  ~WeightedRoundRobin() override;
+
+  void ShutdownLocked() override;
+
+  RefCountedPtr<AddressWeight> GetOrCreateWeight(
+      const grpc_resolved_address& address);
+
+  RefCountedPtr<WeightedRoundRobinConfig> config_;
+
+  // List of subchannels.
+  RefCountedPtr<WeightedRoundRobinSubchannelList> subchannel_list_;
+  // Latest pending subchannel list.
+  // When we get an updated address list, we create a new subchannel list
+  // for it here, and we wait to swap it into subchannel_list_ until the new
+  // list becomes READY.
+  RefCountedPtr<WeightedRoundRobinSubchannelList>
+      latest_pending_subchannel_list_;
+
+  Mutex address_weight_map_mu_;
+  std::map<std::string, AddressWeight*, std::less<>> address_weight_map_
+      ABSL_GUARDED_BY(&address_weight_map_mu_);
+
+  bool shutdown_ = false;
+
+  absl::BitGen bit_gen_;
+};
+
+//
+// WeightedRoundRobin::AddressWeight
+//
+
+WeightedRoundRobin::AddressWeight::~AddressWeight() {
+  MutexLock lock(&wrr_->address_weight_map_mu_);
+  auto it = wrr_->address_weight_map_.find(key_);
+  if (it != wrr_->address_weight_map_.end() && it->second == this) {
+    wrr_->address_weight_map_.erase(it);
+  }
+}
+
+void WeightedRoundRobin::AddressWeight::MaybeUpdateWeight(
+    double qps, double cpu_utilization) {
+  // Compute weight.
+  float weight = 0;
+  if (qps > 0 && cpu_utilization > 0) weight = qps / cpu_utilization;
+  if (weight == 0) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_lb_wrr_trace)) {
+      gpr_log(GPR_INFO,
+              "[WRR %p] subchannel %s: qps=%f, cpu_utilization=%f: weight=%f "
+              "(not updating)",
+              wrr_.get(), key_.c_str(), qps, cpu_utilization, weight);
+    }
+    return;
+  }
+  Timestamp now = Timestamp::Now();
+  // Grab the lock and update the data.
+  MutexLock lock(&mu_);
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_lb_wrr_trace)) {
+    gpr_log(GPR_INFO,
+            "[WRR %p] subchannel %s: qps=%f, cpu_utilization=%f: setting "
+            "weight=%f weight_=%f now=%s last_update_time_=%s "
+            "non_empty_since_=%s",
+            wrr_.get(), key_.c_str(), qps, cpu_utilization, weight, weight_,
+            now.ToString().c_str(), last_update_time_.ToString().c_str(),
+            non_empty_since_.ToString().c_str());
+  }
+  if (weight > 0 && non_empty_since_ == Timestamp::InfFuture()) {
+    non_empty_since_ = now;
```

Comment:
We should reset this if the weight is zero as well.

Reply:
If the weight is 0, we return on line 388.

I don't think we want to differentiate between receiveing a backend metric report with all values 0 and not receiving any backend metric report.  I think we want to treat both cases the same by applying the expiration time.

Reply:
I meant to reset the blackout timer before returning.

This is a non-trivial deviation from what the blackout period does and I think we should do what's proven as discussed. And resetting makes sense to me. An explicit zero weight can happen e.g. when the QPS counter is reset/restarted, CPU% going out of range etc. These seem to be as good signals or better to hold off accepting load reports from that subchannel until it stabilizes, as we do to a subchannel comes back online to prevent churns.

Reply:
It looks like you're bouncing back and forth between comment threads here, and it's fairly confusing.  I think I addressed this in the [thread below](https://github.com/grpc/grpc/pull/31904#discussion_r1071639216).